src/browser.ts

```diff
@@ -0,0 +1,214 @@
+import {Descriptor} from './descriptor';
+import {createApiCall as _createApiCall} from './createApiCall';
+
+export {PathTemplate} from './pathTemplate';
+export {CallSettings, constructSettings, RetryOptions} from './gax';
+
+export {
+  BundleDescriptor,
+  LongrunningDescriptor,
+  PageDescriptor,
+  StreamDescriptor,
+} from './descriptor';
+
+export {StreamType} from './streamingCalls/streaming';
+
+export const ALL_SCOPES: string[] = [];
+lro.ALL_SCOPES = ALL_SCOPES;
+
+interface CancelHandler {
+  canceller: AbortController;
+  cancelRequested: boolean;
+}
+
+export class GrpcClient {
+  auth: GoogleAuth;
+  promise?: PromiseConstructor;
+
+  constructor(options: GrpcClientOptions = {}) {
+    if (!options.auth) {
+      throw new Error(
+        'You need to pass auth instance to gRPC-fallback client. Use OAuth2Client from google-auth-library.'
+      );
+    }
+    this.auth = options.auth;
+    this.promise = 'promise' in options ? options.promise! : Promise;
+  }
+
+  loadProto(jsonObject) {
+    const rootObject = protobuf.Root.fromJSON(jsonObject);
+    return rootObject;
+  }
+
+  private getServiceMethods(service: protobuf.Service) {
+    const methods = Object.keys(service.methods);
+
+    // Creating an array of all methods within the stub with correct case (i.e, first letter is lower-case)
+    const methodsCorrectCase = methods.map(method => {
+      return method[0].toLowerCase() + method.substring(1);
+    });
+
+    return methodsCorrectCase;
+  }
+
+  constructSettings(
+    serviceName: string,
+    clientConfig: gax.ClientConfig,
+    configOverrides: gax.ClientConfig,
+    headers: OutgoingHttpHeaders
+  ) {
+    return gax.constructSettings(
+      serviceName,
+      clientConfig,
+      configOverrides,
+      Status,
+      this.promise
+    );
+  }
+
+  async createStub(service: protobuf.Service, opts: ClientStubOptions) {
+    const authHeader = await this.auth.getRequestHeaders();
+    function serviceClientImpl(method, requestData, callback) {
+      let cancelController, cancelSignal;
+      if (typeof AbortController !== 'undefined') {
+        cancelController = new AbortController();
+        cancelSignal = cancelController.signal;
+      }
+      const cancelHandler: CancelHandler = {
+        canceller: cancelController,
+        cancelRequested: false,
+      };
+      const headers = Object.assign({}, authHeader);
+      headers['Content-Type'] = 'application/x-protobuf';
+      headers['User-Agent'] = 'testapp/1.0';
+
+      const grpcFallbackProtocol = opts.protocol || 'https';
+      let servicePath = opts.servicePath;
+      if (!servicePath) {
+        if (service.options && service.options['(google.api.default_host)']) {
+          servicePath = service.options['(google.api.default_host)'];
+        } else {
+          callback(new Error('Service path is undefined'));
+          return;
+        }
+      }
+      const servicePort = opts.port || 443;
+      const serviceNameArray: string[] = [];
```

Comment:
by `serviceName` here do we mean the fully-qualified name of the proto Service? If so, I think a comment or better naming should be used here and the relevant code following. The word "service" is pretty overloaded (even in our own code).

Reply:
Look at the `url` below, it has multiple uses of the word `service` in its construction and they mean different things...

Reply:
+1 here, `serviceNameArray` relates to protobuf service name (`google.cloud.speech.v1.Speech`) and in the same code we have `servicePath` which is the hostname. Let's figure out the better naming here.

Reply:
In the grpc-fallback section of [this link](https://googleapis.github.io/HowToRPC) it describes the `serviceName` part as the fully qualified protobuf service name, which is the package name + name of the service section of the proto. Would something like `fullServicePackageName` or 'protoPackageName` make sense here or is that ambiguous/misleading?

Reply:
That's probably not a _package_ name, it's still a _service_ name. Maybe `protoServiceName` (for the resulting string, e.g. `google.cloud.speech.v1.Speech`) and `protoNamespaces` for an array (`['google', 'cloud', 'speech', 'v1'])?  @noahdietz what do you think?